client/network/src/transport.rs

```diff
@@ -42,14 +42,22 @@ pub fn build_transport(
 ) -> (Boxed<(PeerId, StreamMuxerBox), io::Error>, Arc<bandwidth::BandwidthSinks>) {
 	// Build configuration objects for encryption mechanisms.
 	let noise_config = {
-		let noise_keypair = noise::Keypair::<noise::X25519>::new().into_authentic(&keypair)
-			// For more information about this panic, see in "On the Importance of Checking
-			// Cryptographic Protocols for Faults" by Dan Boneh, Richard A. DeMillo,
-			// and Richard J. Lipton.
+		// For more information about these two panics, see in "On the Importance of
+		// Checking Cryptographic Protocols for Faults" by Dan Boneh, Richard A. DeMillo,
+		// and Richard J. Lipton.
+		let noise_keypair_legacy = noise::Keypair::<noise::X25519>::new().into_authentic(&keypair)
 			.expect("can only fail in case of a hardware bug; since this signing is performed only \
 				once and at initialization, we're taking the bet that the inconvenience of a very \
 				rare panic here is basically zero");
-		noise::NoiseConfig::ix(noise_keypair)
+		let noise_keypair_spec = noise::Keypair::<noise::X25519Spec>::new().into_authentic(&keypair)
+			.expect("can only fail in case of a hardware bug; since this signing is performed only \
+				once and at initialization, we're taking the bet that the inconvenience of a very \
+				rare panic here is basically zero");
+
+		core::upgrade::SelectUpgrade::new(
+			noise::NoiseConfig::ix(noise_keypair_legacy),
+			noise::NoiseConfig::xx(noise_keypair_spec)
```

Comment:
Interesting, I didn't realise we used `ix`. Strictly speaking there is actually no need for substrate to use `xx`, since `X25519Spec` still supports `ix`, unless compatibility with other libp2p implementations is a concern. In any case, shouldn't we switch the order so that the `X25519Spec` negotiation is tried first? Otherwise it won't actually be used during the transition period and we may be in for a surprise if the legacy negotiation is eventually removed but the other doesn't work for some reason. Switching the order would also mean that upgraded nodes will already prefer the non-legacy protocol.

Reply:
I've switched to `xx` here because it's actually the only handshake that it is possible to use.
The upgrade trait is only implemented for `XX`: https://github.com/libp2p/rust-libp2p/blob/bd5e81fedbc5aa9bf941a8ba5da02dee26f291e3/protocols/noise/src/protocol/x25519_spec.rs#L72-L79
While for the legacy protocol, it is implemented on all three: https://github.com/libp2p/rust-libp2p/blob/bd5e81fedbc5aa9bf941a8ba5da02dee26f291e3/protocols/noise/src/protocol/x25519.rs#L71-L96

> shouldn't we switch the order so that the X25519Spec negotiation is tried first?

Indeed! I actually wanted to do that and then forgot.

> unless compatibility with other libp2p implementations is a concern.

Ideally we would indeed be compatible with the other libp2p implementations.

One option could be to support `ix` by default, and `xx` as the fallback, both using the spec-compliant protocol.

Reply:
> I've switched to `xx` here because it's actually the only handshake that it is possible to use.
> The upgrade trait is only implemented for `XX`: https://github.com/libp2p/rust-libp2p/blob/bd5e81fedbc5aa9bf941a8ba5da02dee26f291e3/protocols/noise/src/protocol/x25519_spec.rs#L72-L79
> While for the legacy protocol, it is implemented on all three: https://github.com/libp2p/rust-libp2p/blob/bd5e81fedbc5aa9bf941a8ba5da02dee26f291e3/protocols/noise/src/protocol/x25519.rs#L71-L96

Oops, that was an oversight.

> Ideally we would indeed be compatible with the other libp2p implementations.
> 
> One option could be to support `ix` by default, and `xx` as the fallback, both using the spec-compliant protocol.

I don't really think saving the 0.5RTT is worth it, especially if you then add another negotiation pass with nodes that don't speak `ix`. If compatibility is desired, `xx` is the way to go.